### DISCUSSION TITLE
RP2350: Entering a floating point value directly

### DISCUSSION THREAD

**Discussion by mcctuxic** [tabemann/zeptoforth#162] [Q&A]
Hi,

yesterday I finally received my Raspberry Pi Pico 2 W with a RP2350 on it! :)

Shortly after I flashed Zeptoforth 1.9.0 to it...

To enter A 15.16 fixed point value I enter (for example):

3;141599

then one can convert this to a floating point value with f32>v.

But how can I enter a floating point value directly and save the conversion step?
Or am just too tired this evening to recognize the obvious? ;)

I looked into .../docs/words/float32.html but didn't find a hint...

Cheers!
mcc

**Comment by tabemann** [comment]
On Wed, Dec 4, 2024 at 11:35 AM mcctuxic ***@***.***> wrote:

> Hi,
>
> yesterday I finally received my Raspberry Pi Pico 2 W with a RP2350 on it!
> :)
>
> Shortly after I flashed Zeptoforth 1.9.0 to it...
>

Warning -- it turns out there is a bug in parsing double-cell values in
zeptoforth 1.9.0. Also, the CYW43439 driver used by zeptoIP has a bug on
the RP2350 where it oftentimes will fail to talk to the CYW43439 chip, and
I have seen very hard crashes occur later if connection is initially
successful. (I have fixes in the devel and master branches for these
issues, but I have not made a release incorporating them yet.)

Another note, the pesky USB CDC problems often prevent uploading the
CYW43439 driver and zeptoIP to the Raspberry Pi Pico 2 W. There is a
workaround, which is to issue INT-IO::SERIAL-CONSOLE and then upload them
with a serial link. There is a new USB CDC console in the works which is
supposed to fix these problems, but it clearly does not work fully since it
too locked up hard in the midst of a mass upload of zeptoIP.

To enter A 15.16 fixed point value I enter (for example):
>
> 3;141599
>
> then one can convert this to a floating point value with f32>v.
>
> But how can I enter a floating point value directly and save the
> conversion step?
> Or am just too tired this evening to recognize the obvious? ;)
>
> I looked into .../docs/words/float32.html but didn't find a hint...
>
> Cheers!
> mcc
>

The numeric format used for single-precision floating-point values would be
3.141599e0 -- do note that the 'e' is mandatory, because any number that is
just (-)x.y without an e is treated as a normal double-cell value. Also
note that these numbers are always decimal, even if BASE is set to
something other than 10.

Travis

**Comment by mcctuxic** [comment]
On 12/04 01:43, tabemann wrote:
> On Wed, Dec 4, 2024 at 11:35 AM mcctuxic ***@***.***> wrote:
> 
> > Hi,
> >
> > yesterday I finally received my Raspberry Pi Pico 2 W with a RP2350 on it!
> > :)
> >
> > Shortly after I flashed Zeptoforth 1.9.0 to it...
> >
> 
> Warning -- it turns out there is a bug in parsing double-cell values in
> zeptoforth 1.9.0. Also, the CYW43439 driver used by zeptoIP has a bug on
> the RP2350 where it oftentimes will fail to talk to the CYW43439 chip, and
> I have seen very hard crashes occur later if connection is initially
> successful. (I have fixes in the devel and master branches for these
> issues, but I have not made a release incorporating them yet.)
> 
> Another note, the pesky USB CDC problems often prevent uploading the
> CYW43439 driver and zeptoIP to the Raspberry Pi Pico 2 W. There is a
> workaround, which is to issue INT-IO::SERIAL-CONSOLE and then upload them
> with a serial link. There is a new USB CDC console in the works which is
> supposed to fix these problems, but it clearly does not work fully since it
> too locked up hard in the midst of a mass upload of zeptoIP.
> 
> To enter A 15.16 fixed point value I enter (for example):
> >
> > 3;141599
> >
> > then one can convert this to a floating point value with f32>v.
> >
> > But how can I enter a floating point value directly and save the
> > conversion step?
> > Or am just too tired this evening to recognize the obvious? ;)
> >
> > I looked into .../docs/words/float32.html but didn't find a hint...
> >
> > Cheers!
> > mcc
> >
> 
> The numeric format used for single-precision floating-point values would be
> 3.141599e0 -- do note that the 'e' is mandatory, because any number that is
> just (-)x.y without an e is treated as a normal double-cell value. Also
> note that these numbers are always decimal, even if BASE is set to
> something other than 10.
> 
> Travis
> 
> > Message ID: ***@***.***>
> >


Hi Travis,

thank you for the bonus info! :)

I am using GENTOO linux, wezterm 20240203-110809-5046fc22-gentoo and
picocom 3.1

I got this (copy of the console output):
picocom v3.1

port is        : /dev/ttyACM0
flowcontrol    : xon/xoff
baudrate is    : 115200
parity is      : none
databits are   : 8
stopbits are   : 1
escape is      : C-a
local echo is  : no
noinit is      : no
noreset is     : no
hangup is      : no
nolock is      : no
send_cmd is    : sz -vv
receive_cmd is : rz -vv -E
imap is        :
omap is        :
emap is        : crcrlf,delbs,
logfile is     : none
initstring     : none
exit_after is  : not set
exit is        : no

Type [C-a] [C-h] to see available commands
Terminal ready

Welcome to zeptoforth
Built for rp2350, version 1.9.0, on Thu Nov 14 01:33:19 PM CST 2024
zeptoforth comes with ABSOLUTELY NO WARRANTY: for details type `license'
 ok
  ok
float32 import  ok
3.141599e0 unable to parse: 3.141599e0


Seems not to work...

Cheers!
mcc

**Comment by tabemann** [comment]
On Wed, Dec 4, 2024 at 8:36 PM mcctuxic ***@***.***> wrote:

> On 12/04 01:43, tabemann wrote:
> > On Wed, Dec 4, 2024 at 11:35 AM mcctuxic ***@***.***> wrote:
> >
> > > Hi,
> > >
> > > yesterday I finally received my Raspberry Pi Pico 2 W with a RP2350 on
> it!
> > > :)
> > >
> > > Shortly after I flashed Zeptoforth 1.9.0 to it...
> > >
> >
> > Warning -- it turns out there is a bug in parsing double-cell values in
> > zeptoforth 1.9.0. Also, the CYW43439 driver used by zeptoIP has a bug on
> > the RP2350 where it oftentimes will fail to talk to the CYW43439 chip,
> and
> > I have seen very hard crashes occur later if connection is initially
> > successful. (I have fixes in the devel and master branches for these
> > issues, but I have not made a release incorporating them yet.)
> >
> > Another note, the pesky USB CDC problems often prevent uploading the
> > CYW43439 driver and zeptoIP to the Raspberry Pi Pico 2 W. There is a
> > workaround, which is to issue INT-IO::SERIAL-CONSOLE and then upload them
> > with a serial link. There is a new USB CDC console in the works which is
> > supposed to fix these problems, but it clearly does not work fully since
> it
> > too locked up hard in the midst of a mass upload of zeptoIP.
> >
> > To enter A 15.16 fixed point value I enter (for example):
> > >
> > > 3;141599
> > >
> > > then one can convert this to a floating point value with f32>v.
> > >
> > > But how can I enter a floating point value directly and save the
> > > conversion step?
> > > Or am just too tired this evening to recognize the obvious? ;)
> > >
> > > I looked into .../docs/words/float32.html but didn't find a hint...
> > >
> > > Cheers!
> > > mcc
> > >
> >
> > The numeric format used for single-precision floating-point values would
> be
> > 3.141599e0 -- do note that the 'e' is mandatory, because any number that
> is
> > just (-)x.y without an e is treated as a normal double-cell value. Also
> > note that these numbers are always decimal, even if BASE is set to
> > something other than 10.
> >
> > Travis
> >
> > > Message ID: ***@***.***>
> > >
>
>
> Hi Travis,
>
> thank you for the bonus info! :)
>
> I am using GENTOO linux, wezterm 20240203-110809-5046fc22-gentoo and
> picocom 3.1
>
> I got this (copy of the console output):
> picocom v3.1
>
> port is : /dev/ttyACM0
> flowcontrol : xon/xoff
> baudrate is : 115200
> parity is : none
> databits are : 8
> stopbits are : 1
> escape is : C-a
> local echo is : no
> noinit is : no
> noreset is : no
> hangup is : no
> nolock is : no
> send_cmd is : sz -vv
> receive_cmd is : rz -vv -E
> imap is :
> omap is :
> emap is : crcrlf,delbs,
> logfile is : none
> initstring : none
> exit_after is : not set
> exit is : no
>
> Type [C-a] [C-h] to see available commands
> Terminal ready
>
> Welcome to zeptoforth
> Built for rp2350, version 1.9.0, on Thu Nov 14 01:33:19 PM CST 2024
> zeptoforth comes with ABSOLUTELY NO WARRANTY: for details type `license'
> ok
> ok
> float32 import ok
> 3.141599e0 unable to parse: 3.141599e0
>
>
> Seems not to work...
>
> Cheers!
> mcc
>
> When I tried to reproduce I got this:

Welcome to zeptoforth
Built for rp2350, version 1.9.0, on Thu Nov 14 01:33:19 PM CST 2024
zeptoforth comes with ABSOLUTELY NO WARRANTY: for details type `license'
 ok
float32 import  ok
3.141599e0 v. 3.141599e0  ok

Technically FLOAT32 is not needed to parse single-precision floating-point
numbers, but it is needed for V. to print them out.

Note that I had just freshly flashed my Pico 2 W with
bin/1.9.0/rp2350/zeptoforth_full_usb-1.9.0.uf2 to make sure I was using the
exact code that was in the distribution tarball. This is without any
bugfixes or like added.

Travis

**Comment by mcctuxic** [comment]
On 12/05 07:37, tabemann wrote:
> On Wed, Dec 4, 2024 at 8:36 PM mcctuxic ***@***.***> wrote:
> 
> > On 12/04 01:43, tabemann wrote:
> > > On Wed, Dec 4, 2024 at 11:35 AM mcctuxic ***@***.***> wrote:
> > >
> > > > Hi,
> > > >
> > > > yesterday I finally received my Raspberry Pi Pico 2 W with a RP2350 on
> > it!
> > > > :)
> > > >
> > > > Shortly after I flashed Zeptoforth 1.9.0 to it...
> > > >
> > >
> > > Warning -- it turns out there is a bug in parsing double-cell values in
> > > zeptoforth 1.9.0. Also, the CYW43439 driver used by zeptoIP has a bug on
> > > the RP2350 where it oftentimes will fail to talk to the CYW43439 chip,
> > and
> > > I have seen very hard crashes occur later if connection is initially
> > > successful. (I have fixes in the devel and master branches for these
> > > issues, but I have not made a release incorporating them yet.)
> > >
> > > Another note, the pesky USB CDC problems often prevent uploading the
> > > CYW43439 driver and zeptoIP to the Raspberry Pi Pico 2 W. There is a
> > > workaround, which is to issue INT-IO::SERIAL-CONSOLE and then upload them
> > > with a serial link. There is a new USB CDC console in the works which is
> > > supposed to fix these problems, but it clearly does not work fully since
> > it
> > > too locked up hard in the midst of a mass upload of zeptoIP.
> > >
> > > To enter A 15.16 fixed point value I enter (for example):
> > > >
> > > > 3;141599
> > > >
> > > > then one can convert this to a floating point value with f32>v.
> > > >
> > > > But how can I enter a floating point value directly and save the
> > > > conversion step?
> > > > Or am just too tired this evening to recognize the obvious? ;)
> > > >
> > > > I looked into .../docs/words/float32.html but didn't find a hint...
> > > >
> > > > Cheers!
> > > > mcc
> > > >
> > >
> > > The numeric format used for single-precision floating-point values would
> > be
> > > 3.141599e0 -- do note that the 'e' is mandatory, because any number that
> > is
> > > just (-)x.y without an e is treated as a normal double-cell value. Also
> > > note that these numbers are always decimal, even if BASE is set to
> > > something other than 10.
> > >
> > > Travis
> > >
> > > > Message ID: ***@***.***>
> > > >
> >
> >
> > Hi Travis,
> >
> > thank you for the bonus info! :)
> >
> > I am using GENTOO linux, wezterm 20240203-110809-5046fc22-gentoo and
> > picocom 3.1
> >
> > I got this (copy of the console output):
> > picocom v3.1
> >
> > port is : /dev/ttyACM0
> > flowcontrol : xon/xoff
> > baudrate is : 115200
> > parity is : none
> > databits are : 8
> > stopbits are : 1
> > escape is : C-a
> > local echo is : no
> > noinit is : no
> > noreset is : no
> > hangup is : no
> > nolock is : no
> > send_cmd is : sz -vv
> > receive_cmd is : rz -vv -E
> > imap is :
> > omap is :
> > emap is : crcrlf,delbs,
> > logfile is : none
> > initstring : none
> > exit_after is : not set
> > exit is : no
> >
> > Type [C-a] [C-h] to see available commands
> > Terminal ready
> >
> > Welcome to zeptoforth
> > Built for rp2350, version 1.9.0, on Thu Nov 14 01:33:19 PM CST 2024
> > zeptoforth comes with ABSOLUTELY NO WARRANTY: for details type `license'
> > ok
> > ok
> > float32 import ok
> > 3.141599e0 unable to parse: 3.141599e0
> >
> >
> > Seems not to work...
> >
> > Cheers!
> > mcc
> >
> > When I tried to reproduce I got this:
> 
> Welcome to zeptoforth
> Built for rp2350, version 1.9.0, on Thu Nov 14 01:33:19 PM CST 2024
> zeptoforth comes with ABSOLUTELY NO WARRANTY: for details type `license'
>  ok
> float32 import  ok
> 3.141599e0 v. 3.141599e0  ok
> 
> Technically FLOAT32 is not needed to parse single-precision floating-point
> numbers, but it is needed for V. to print them out.
> 
> Note that I had just freshly flashed my Pico 2 W with
> bin/1.9.0/rp2350/zeptoforth_full_usb-1.9.0.uf2 to make sure I was using the
> exact code that was in the distribution tarball. This is without any
> bugfixes or like added.
> 
> Travis
> 
> > Message ID: <tabemann/zeptoforth/repo-discussions/162/comments/11468113@
> > github.com>
> >
> 

Hi Travis,

I reflashed my Pi Pico 2 W with the uf2-image you used and got the same
result as you did this time:

Welcome to zeptoforth
Built for rp2350, version 1.9.0, on Thu Nov 14 01:33:19 PM CST 2024
zeptoforth comes with ABSOLUTELY NO WARRANTY: for details type `license'
 ok
float32 import  ok
3.141599e0 v. 3.141599e0  ok

Then I flashed zed into flash after adding a marker and re-did the
experiment and it worked as above.

Then I flashed fixed32 to the board...and now the parsing failed
as mentioned in my previous post.

Since there are conversion routines from fixed-format to floating-format
I thought, both could co-exist...

Am I wrong?

Cheers!
mcc

**Comment by tabemann** [comment]
On Thu, Dec 5, 2024 at 11:08 AM mcctuxic ***@***.***> wrote:

> Hi Travis,
>
> I reflashed my Pi Pico 2 W with the uf2-image you used and got the same
> result as you did this time:
>
> Welcome to zeptoforth
> Built for rp2350, version 1.9.0, on Thu Nov 14 01:33:19 PM CST 2024
> zeptoforth comes with ABSOLUTELY NO WARRANTY: for details type `license'
> ok
> float32 import ok
> 3.141599e0 v. 3.141599e0 ok
>
> Then I flashed zed into flash after adding a marker and re-did the
> experiment and it worked as above.
>
> Then I flashed fixed32 to the board...and now the parsing failed
> as mentioned in my previous post.
>
> Since there are conversion routines from fixed-format to floating-format
> I thought, both could co-exist...
>
> Am I wrong?
>
> Cheers!
> mcc
>

Thanks for finding this! That points to a bug in FIXED32, because it should
not be behaving that way. I will look into making a fix to it today.
Hopefully I can get a new bugfix release out over the weekend.

Travis

**Comment by mcctuxic** [comment]
On 12/05 09:32, tabemann wrote:
> On Thu, Dec 5, 2024 at 11:08 AM mcctuxic ***@***.***> wrote:
> 
> > Hi Travis,
> >
> > I reflashed my Pi Pico 2 W with the uf2-image you used and got the same
> > result as you did this time:
> >
> > Welcome to zeptoforth
> > Built for rp2350, version 1.9.0, on Thu Nov 14 01:33:19 PM CST 2024
> > zeptoforth comes with ABSOLUTELY NO WARRANTY: for details type `license'
> > ok
> > float32 import ok
> > 3.141599e0 v. 3.141599e0 ok
> >
> > Then I flashed zed into flash after adding a marker and re-did the
> > experiment and it worked as above.
> >
> > Then I flashed fixed32 to the board...and now the parsing failed
> > as mentioned in my previous post.
> >
> > Since there are conversion routines from fixed-format to floating-format
> > I thought, both could co-exist...
> >
> > Am I wrong?
> >
> > Cheers!
> > mcc
> >
> 
> Thanks for finding this! That points to a bug in FIXED32, because it should
> not be behaving that way. I will look into making a fix to it today.
> Hopefully I can get a new bugfix release out over the weekend.
> 
> Travis
> 
> > Message ID: <tabemann/zeptoforth/repo-discussions/162/comments/11475957@
> > github.com>
> >

Hi Travis,

oh! :)
But I didn't want to put any pressure - neither implicit nor explicit -
on you!!!!!!! Weekend is weekend even for a Forth guru like you :::)))

Thank you very much for Zeptoforth! 

Cheers and have a good <insert time in your timezone here> ;)
mcc

**Comment by tabemann** [comment]
On Thu, Dec 5, 2024 at 11:44 AM mcctuxic ***@***.***> wrote:

> Hi Travis,
>
> oh! :)
> But I didn't want to put any pressure - neither implicit nor explicit -
> on you!!!!!!! Weekend is weekend even for a Forth guru like you :::)))
>
> Thank you very much for Zeptoforth!
>
> Cheers and have a good <insert time in your timezone here> ;)
> mcc
>
> I've made a fix to this issue, which was actually trivial ─ it just turned
out that I was directly calling an earlier instance of DO-HANDLE-NUMBER
rather than saving HANDLE-NUMBER-HOOK and then executing the saved
HANDLE-NUMBER-HOOK . I don't like having outstanding unfixed bugs in
releases which is why I'll try to put out a release when I get a chance
(and this weekend is when I'll have time to do so).

Travis

**Comment by mcctuxic** [comment]
On 12/05 10:08, tabemann wrote:
> On Thu, Dec 5, 2024 at 11:44 AM mcctuxic ***@***.***> wrote:
> 
> > Hi Travis,
> >
> > oh! :)
> > But I didn't want to put any pressure - neither implicit nor explicit -
> > on you!!!!!!! Weekend is weekend even for a Forth guru like you :::)))
> >
> > Thank you very much for Zeptoforth!
> >
> > Cheers and have a good <insert time in your timezone here> ;)
> > mcc
> >
> > I've made a fix to this issue, which was actually trivial ─ it just turned
> out that I was directly calling an earlier instance of DO-HANDLE-NUMBER
> rather than saving HANDLE-NUMBER-HOOK and then executing the saved
> HANDLE-NUMBER-HOOK . I don't like having outstanding unfixed bugs in
> releases which is why I'll try to put out a release when I get a chance
> (and this weekend is when I'll have time to do so).
> 
> Travis
> 
> > Message ID: <tabemann/zeptoforth/repo-discussions/162/comments/11476302@
> > github.com>
> >
> 

Hi Travis,

a bug fix in the range of a view hours including time zone
differences...whow!
It seems, not only FORTH is fast <grin> :)

As always: Thanks a lot!

Cheers!
mcc
